pyro/autohelp/autohelp.py

```diff
@@ -128,86 +153,104 @@ async def process_message(
         if key in self._help_cache:
             return
 
-        code_bin_content = await self._code_bin.process(message.content)
-        message.content += code_bin_content
-        message.content = message.content = message.content.replace("\r", "")
+        contents = await self.find_code(message)
+        if not contents:
+            return
+        content = contents[0]
+        # prelim check
+        # TODO: implement prelimary check and iterate over all provided code snippets
 
-        iters = [call(message) for call in self.patterns]
-        results = await asyncio.gather(*iters)
-        results = list(filter(None, results))
-        if not results:
+        # parse the contents
+        try:
+            parsed = libcst.parse_module(content)
+        except libcst.ParserSyntaxError:
+            return
+        results = parsed
+        visitors: list[libcst.CSTTransformer] = []
+        errors = []
+        for visitor in self.visitors:
+            visitor = visitor()
+            visitors.append(visitor)
+            results = results.visit(visitor)
+            if visitor.found_errors:
+                errors.extend(visitor.found_errors)
+
+        if not errors:
             return None
 
-        self._help_cache.add_entry(key, None, ttl=datetime.timedelta(minutes=30))
-
-        auto_message = await message.channel.send(
-            f"{message.author.mention} {'this' if len(results) == 1 else 'these'} might help.",
-            embeds=results,
-        )
+        errors = set(errors)
+        actions = []
+        original_code = parsed.code
+        new_code = results.code
+
+        original_code_split = original_code.splitlines(keepends=True)
+        new_code_split = new_code.splitlines(keepends=True)
+        original_code = []
+        new_code = []
+
+        for diff in difflib.unified_diff(original_code_split, new_code_split, n=1):
+
+            if diff.startswith("@@"):
+                original_code.append("")
+                new_code.append("")
+            elif diff.strip() in ("+++", "---"):
+                continue
+            elif diff.startswith("+"):
+                new_code[-1] += diff[1:]
+            elif diff.startswith("-"):
+                original_code[-1] += diff[1:]
+            elif diff.startswith(" "):
+                original_code[-1] += diff[1:]
+                new_code[-1] += diff[1:]
+
+        for error, action in self.actions.items():
+            if error in errors:
+                actions.append(action(message))
+        embed, last_fields = self.build_embed(message, original_code, new_code)
+        print(actions)
```

Comment:
-> logging

Reply:
This shouldn't even be here, related to figuring out the cause of https://bugs.python.org/issue46672